slave_stack/src/ESC/Lan9252.cc

```diff
@@ -0,0 +1,183 @@
+        {
+            counter++;
+            readInternalRegister(BYTE_TEST, byte_test_result);
+        }
+
+        if (counter == timeout)
+        {
+          Serial.println("Timeout get byte test");
+        }
+        Serial.print("Byte test read: ");
+        Serial.println(byte_test_result, HEX);
+
+
+        uint32_t hw_cfg_ready = 0;
+        counter = 0;
+        while (counter < timeout and not (hw_cfg_ready & DEVICE_READY))
+        {
+            counter++;
+            readInternalRegister(HW_CFG, hw_cfg_ready);
+        }
+
+        if (counter == timeout)
+        {
+          Serial.println("Timeout hw cfg ready");
+        }
+
+
+        spi_interface_.endTransaction();
+    }
+
+
+    int32_t Lan9252::waitCSRReady()
+    {
+        uint32_t esc_status;
+        do
+        {
+            readInternalRegister(ECAT_CSR_CMD, esc_status);
+        }
+        while(esc_status & ECAT_CSR_BUSY);
+        return 0;
+    }
+
+
+    void Lan9252::readInternalRegister(uint16_t address, void* payload, uint32_t size)
+    {
+        InternalRegisterControl cmd{READ, hton(address), {}};
+
+        spi_interface_.enableChipSelect();
+        spi_interface_.write(&cmd, CSR_CMD_HEADER_SIZE);
+
+        spi_interface_.read(payload, size);
+        spi_interface_.disableChipSelect();
+    }
+
+
+    void Lan9252::writeInternalRegister(uint16_t address, void const* payload, uint32_t size)
+    {
+        // TODO check payload size, return code too big ?
+        InternalRegisterControl cmd{WRITE, hton(address), {}};
+        memcpy(cmd.payload, payload, size);
+        spi_interface_.enableChipSelect();
+        spi_interface_.write(&cmd, CSR_CMD_HEADER_SIZE + size);
+        spi_interface_.disableChipSelect();
+    }
+
+
+    int32_t Lan9252::readRegister(uint16_t address, void* data, uint32_t size)
+    {
+        if (address < 0x1000)
+        {
+            writeInternalRegister(ECAT_CSR_CMD, CSR_CMD{address, size, CSR_CMD::ESC_READ});
+            waitCSRReady();
+            readInternalRegister(ECAT_CSR_DATA, data, size);
+        }
+        else if (address < 0x2000)
+        {
+            writeInternalRegister(ECAT_PRAM_RD_CMD, PRAM_ABORT);
+            uint32_t addr_len = address | (size << 16);             // check size alignment and max value.
+            writeInternalRegister(ECAT_PRAM_RD_ADDR_LEN, addr_len);
+            writeInternalRegister(ECAT_PRAM_RD_CMD, PRAM_BUSY);  // order start read
+
+            uint16_t fifo_slot_available; // slot of 4 bytes
+            do
+            {
+                readInternalRegister(ECAT_PRAM_RD_CMD, fifo_slot_available);
+                fifo_slot_available = fifo_slot_available >> 8;
+            } while(fifo_slot_available != size / 4);  // TODO beware assumption ESC will transfer 32 bytes.
+
+            readInternalRegister(ECAT_PRAM_RD_DATA, data, size);
+        }
+
+        return 0;
+    };
+
+    int32_t Lan9252::writeRegister(uint16_t address, void const* data, uint32_t size)
```

Comment:
Rename as read (and not readRegister) because it enable reading the whole memory of the ESC. Same for write command.